File: 388.文件的最长绝对路径.go
/*
 * @lc app=leetcode.cn id=388 lang=golang
 *
 * [388] 文件的最长绝对路径
 */

// @lc code=start
func lengthLongestPath(input string) (res int) {
	nodes := strings.Split(input, "\n")
	length := []int{0}
	for i := range nodes {
		t, l := getTab(nodes[i])
		if t > 0 {
			l += length[t-1]
		}
		if isFile(nodes[i]) {
			res = max(res, l+t)
		}
		if t >= len(length) {
			length = append(length, l)
		} else {
			length[t] = l
		}
	}
	return
}
func isFile(str string) bool {
	return strings.Contains(str, ".")
}
func getTab(str string) (t, l int) {
	for i := range str {
		if str[i] == '\t' {
			t++
		} else {
			break
		}
	}
	return t, len(str) - t
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
// @lc code=end

File: 388.文件的最长绝对路径.py
#
# @lc app=leetcode.cn id=388 lang=python3
#
# [388] 文件的最长绝对路径
#

# @lc code=start
class Solution:
    def lengthLongestPath(self, input: str) -> int:
        st = []
        ans, i, n = 0, 0, len(input)
        while i < n:
            # 检测当前文件的深度
            depth = 1
            while i < n and input[i] == '\t':
                depth += 1
                i += 1

            # 统计当前文件名的长度
            length, isFile = 0, False
            while i < n and input[i] != '\n':
                if input[i] == '.':
                    isFile = True
                length += 1
                i += 1
            i += 1  # 跳过换行符

            while len(st) >= depth:
                st.pop()
            if st:
                length += st[-1] + 1
            if isFile:
                ans = max(ans, length)
            else:
                st.append(length)
        return ans
# @lc code=end

